java/arcs/core/storage/StorageProxy.kt

```diff
@@ -37,57 +37,50 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperation, T>(
     initialCrdt: CrdtModel<Data, Op, T>
 ) {
     private val log = TaggedLog { "StorageProxy" }
-    private val mutex = reentrantLock()
+    private val handlesMutex = Mutex()
     private val handles: MutableSet<Handle<Data, Op, T>> = mutableSetOf()
+    private var hasReaders = false
+    private val syncMutex = Mutex()
     private val crdt = initialCrdt
     private val waitingSyncs: MutableList<CompletableDeferred<ValueAndVersion<T>>> = mutableListOf()
     private var synchronized = false
-    private var listenerAttached = false
-    private var hasReaders = false
 
     private val store = storeEndpointProvider.getStorageEndpoint()
 
+    init {
+        store.setCallback(ProxyCallback(::onMessage))
+    }
+
     /**
      * Connects a handle. If the handle is readable, it will receive the configured callbacks.
      */
     suspend fun registerHandle(handle: Handle<Data, Op, T>): VersionMap {
         // non-readers don't get callbacks, return early
-        if (!handle.canRead) return mutex.withLock { crdt.versionMap.copy() }
+        if (!handle.canRead) return run { crdt.versionMap.copy() }
```

Comment:
why the use of `run` here?

Reply:
I think because run / let will both return the closure result.
Doesn't this access to the CRDT need to be under lock though?

Reply:
Right, but why the need for a closure at all, for a single expression? 

I have the same question as you about locking, as well...

Reply:
ooops, good catch! Thanks!